src/ast.rs

```diff
@@ -311,6 +350,37 @@ impl<'a> Program<'a> {
     }
 }
 
+#[derive(Debug)]
+struct ProgramBuilderState {
```

Comment:
Hm, I would expect a `struct ProgramBuilder` and a `impl ProgramBuilder { def build ... }`. What do you think?